pkg/client/client.go

```diff
@@ -37,66 +36,139 @@ import (
 
 const (
 	defaultEndpointPort = "9440"
-	ProviderName        = "nutanix"
-	configPath          = "/etc/nutanix/config"
-	endpointKey         = "prismCentral"
-	capxNamespaceKey    = "POD_NAMESPACE"
+
+	capxNamespaceKey = "POD_NAMESPACE"
+)
+
+var (
+	// use a var to override in tests
+	configPath = "/etc/nutanix/config/prismCentral"
 )
 
-var ErrPrismAddressNotSet = fmt.Errorf("cannot get credentials if Prism Address is not set")
-var ErrPrismPortNotSet = fmt.Errorf("cannot get credentials if Prism Port is not set")
+var (
+	ErrPrismAddressNotSet = fmt.Errorf("cannot get credentials if Prism Address is not set")
+	ErrPrismPortNotSet    = fmt.Errorf("cannot get credentials if Prism Port is not set")
+
+	ErrPrismIUsernameNotSet = fmt.Errorf("could not create client because username was not set")
+	ErrPrismIPasswordNotSet = fmt.Errorf("could not create client because password was not set")
+
+	ErrCredentialRefNotSet = fmt.Errorf("credentialRef must be set on CAPX manager")
+)
 
 type NutanixClientHelper struct {
 	secretInformer    coreinformers.SecretInformer
 	configMapInformer coreinformers.ConfigMapInformer
+
+	managerNutanixPrismEndpointReader func() (*credentials.NutanixPrismEndpoint, error)
 }
 
-func NewNutanixClientHelper(secretInformer coreinformers.SecretInformer, cmInformer coreinformers.ConfigMapInformer) (*NutanixClientHelper, error) {
+func NewHelper(secretInformer coreinformers.SecretInformer, cmInformer coreinformers.ConfigMapInformer) *NutanixClientHelper {
 	return &NutanixClientHelper{
-		secretInformer:    secretInformer,
-		configMapInformer: cmInformer,
-	}, nil
+		secretInformer:                    secretInformer,
+		configMapInformer:                 cmInformer,
+		managerNutanixPrismEndpointReader: readManagerNutanixPrismEndpointFromFile,
+	}
+}
+
+func (n *NutanixClientHelper) withCustomNutanixPrismEndpointReader(getter func() (*credentials.NutanixPrismEndpoint, error)) *NutanixClientHelper {
+	n.managerNutanixPrismEndpointReader = getter
+	return n
 }
 
-func (n *NutanixClientHelper) GetClientFromEnvironment(ctx context.Context, nutanixCluster *infrav1.NutanixCluster) (*nutanixClientV3.Client, error) {
+// BuildClientForNutanixClusterWithFallback builds a Nutanix Client from the information provided in nutanixCluster.
+func (n *NutanixClientHelper) BuildClientForNutanixClusterWithFallback(ctx context.Context, nutanixCluster *infrav1.NutanixCluster) (*nutanixClientV3.Client, error) {
+	me, err := n.buildManagementEndpoint(ctx, nutanixCluster)
```

Comment:
Pulled this logic out into its own function because testing `BuildClientForNutanixClusterWithFallback` is complicated since there is nothing really to compare `nutanixClientV3.Client` to as it has an interface and unexorted types in the struct.